File: skills/nodejs-scaffold/scripts/templates/files/src/routes/health.ts
import { readFileSync } from 'node:fs'
import type { FastifyInstance } from 'fastify'
import { checkConnection } from '../db/client.js'

let openapiSpec: string | undefined
try {
  openapiSpec = readFileSync(new URL('../../openapi.yaml', import.meta.url), 'utf8')
} catch {
  openapiSpec = undefined
}

export async function healthRoutes(app: FastifyInstance): Promise<void> {
  app.get('/healthz', async () => 'ok')

  app.get('/readyz', async (_request, reply) => {
    const ok = await checkConnection()
    if (!ok) {
      reply.code(503)
      return 'db unavailable'
    }
    return 'ready'
  })

  app.get('/openapi.yaml', async (_request, reply) => {
    if (!openapiSpec) {
      reply.code(404)
      return
    }
    reply.type('application/yaml')
    return openapiSpec
  })

  app.get('/docs', async (_request, reply) => {
    reply.type('text/html')
    return docsHTML
  })
}

const docsHTML = `<!doctype html>
<html>
  <head>
    <title>API Docs</title>
    <link rel="stylesheet" href="https://unpkg.com/swagger-ui-dist@5/swagger-ui.css" />
  </head>
  <body>
    <div id="swagger-ui"></div>
    <script src="https://unpkg.com/swagger-ui-dist@5/swagger-ui-bundle.js"></script>
    <script>
      window.onload = () => {
        window.ui = SwaggerUIBundle({ url: '/openapi.yaml', dom_id: '#swagger-ui' })
      }
    </script>
  </body>
</html>`

File: skills/go-scaffold/scripts/templates/files/internal/app/health.go
package app

import (
	"context"
	"net/http"
	"time"

	"github.com/jackc/pgx/v5/pgxpool"
)

// Liveness: is the process up. Unauthenticated, and (per app.go) excluded from
// rate limiting.
func handleLiveness(w http.ResponseWriter, _ *http.Request) {
	w.WriteHeader(http.StatusOK)
	_, _ = w.Write([]byte("ok"))
}

// Readiness: can it serve traffic (DB reachable). 503 when the pool can't ping.
func handleReadiness(pool *pgxpool.Pool) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		ctx, cancel := context.WithTimeout(r.Context(), 2*time.Second)
		defer cancel()
		if pool == nil || pool.Ping(ctx) != nil {
			w.WriteHeader(http.StatusServiceUnavailable)
			_, _ = w.Write([]byte("db unavailable"))
			return
		}
		w.WriteHeader(http.StatusOK)
		_, _ = w.Write([]byte("ready"))
	}
}

func handleOpenAPISpec(spec []byte) http.HandlerFunc {
	return func(w http.ResponseWriter, _ *http.Request) {
		w.Header().Set("Content-Type", "application/yaml")
		_, _ = w.Write(spec)
	}
}

func handleDocsUI(w http.ResponseWriter, _ *http.Request) {
	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	_, _ = w.Write([]byte(docsHTML))
}

const docsHTML = `<!doctype html>
<html>
  <head>
    <title>API Docs</title>
    <link rel="stylesheet" href="https://unpkg.com/swagger-ui-dist@5/swagger-ui.css" />
  </head>
  <body>
    <div id="swagger-ui"></div>
    <script src="https://unpkg.com/swagger-ui-dist@5/swagger-ui-bundle.js"></script>
    <script>
      window.onload = () => {
        window.ui = SwaggerUIBundle({ url: '/openapi.yaml', dom_id: '#swagger-ui' })
      }
    </script>
  </body>
</html>`
